docs/MediaCodecTests.md

```diff
@@ -0,0 +1,90 @@
+# How to use MediaCodec Tests
+
+## Background
+
+In Android Media, there is a codec framework team and a video codec quality team. There was a gap in testing for audio codec quality. We wanted to create a tool to find and reproduce audio codec quality issues.
+
+MediaCodec is the Java library used for encoding and decoding video and audio streams. We wanted to focus strictly on audio streams in this project. By using MediaCodec for both encoding and decoding, we can feed audio data through the encoder and the decoder and analyze how much noise is added by the codec.
+
+Below is a diagram of a basic encoding/decoding pipeline.
+
+![MediaCodec basic diagram](images/media-codec-basic-diagram.png "MediaCodec basic diagram")
+
+By varying the types of inputs and encoding parameters, we can better understand the current audio codec quality on Android.
+
+## MediaCodec
+
+MediaCodec can be used to encode data from PCM to OPUS, AAC, AMR-NR, etc. It can also be used to decode data from OPUS, AAC, AMR-NR, etc to PCM. These are considered buffer to buffer as input buffers get packaged to output buffers. Note that multiple input buffers may be used to create an output buffer and vice-versa.
+
+Below is a code block demonstrating how to create a MediaCodec.
+
+![MediaCodec basic code sample](images/media-codec-basic-code-sample.png "MediaCodec basic code sample")
+
+The following steps happen.
+
+1. A specific codec is picked from a list of available codecs.
+2. The codec is configured with a format (sample rate, bitrate, channel count, etc)
+3. The codec is started.
+4. While there is available data to pass into the input buffer, enqueue the buffer.
+5. If there is a buffer available in the output buffer, dequeue the buffer.
+6. Once the data is complete, stop and release the codec.
+
+Below is a diagram on how input buffers turn into output buffers.
+
+![MediaCodec io diagram](images/media-codec-io-diagram.png "MediaCodec io diagram")
+
+For files, the BUFFER_FLAG_END_OF_STREAM flag can be used to signal the end of a stream. By passing this flag when queuing the input buffer, when the same flag is received from the output buffer, you can determine when to terminate the MediaCodec.
+
+## Test Architecture
+
+![MediaCodec full diagram](images/media-codec-full-diagram.png "MediaCodec full diagram")
+
+There are currently 3 ways of testing with MediaCodec
+
+1. A sine wave is created and is passed through a MediaCodec encoder and a MediaCodec decoder before going through a harmonic analyzer. This analyzer will show the total harmonic distortion (THD), total harmonic distortion + noise (THD+N), signal noise ratio (SNR), and the peak decibels. Since a sine wave is a sound only at one frequency, we expect the resulting frequency graph to have one peak. A graph with frequency bins can be looked at to confirm this.
```

Comment:
An FFT can have multiple small side peaks if the sine wave frequency is not centered in a bin.